src/packages/piral-oidc/src/types.ts

```diff
@@ -1,10 +1,91 @@
-declare module 'piral-core/lib/types/custom' {
-  interface PiletCustomApi extends PiralOidcApi {}
+import 'piral-core';
+import { Profile } from 'oidc-client';
+
+/**
+ * Available configuration options for the OpenID Connect plugin.
+ */
+export interface OidcConfig {
+  /**
+   * The id of the client. Required for the setup of OAuth 2.0.
+   */
+  clientId: string;
+  /**
+   * The client secret.
+   */
+  clientSecret?: string;
+  /**
+   * The Uri pointing to the Identity Provider.
+   */
+  identityProviderUri: string;
+  /**
+   * The redirect Uri to use. By default the origin with /auth
+   * is used.
+   */
+  redirectUri?: string;
+  /**
+   * The Uri to which the Identity provider should redirect
+   * after a logout. By default the origin is used.
+   */
+  postLogoutRedirectUri?: string;
+  /**
+   * The protocol response type to be used. By default, `id_token`
+   * is used.
+   */
+  responseType?: string;
+  /**
+   * The scopes to be used. By default, `openid` is used.
+   */
+  scopes?: Array<string>;
+  /**
+   * Restricts token sharing such that other integrations, e.g., with
+   * fetch would need to be done manually.
+   * Otherwise, the client is responsive to the `before-fetch` event.
+   */
+  restrict?: boolean;
 }
 
-export interface PiralOidcApi {
+export type OidcProfileWithCustomClaims<TCustomClaims extends {}> = TCustomClaims & Profile;
+
+export interface OidcRequest {
+  /**
+   * Sets the headers of the request.
+   * @param headers Headers or a promise to headers.
+   */
+  setHeaders(headers: any): void;
+}
+
+export interface OidcClient<TCustomClaims extends {} = {}> {
+  /**
+   * Performs a login.
+   */
+  login(): void;
+  /**
+   * Performs a logout.
+   */
+  logout(): void;
+  /**
+   * Retrieves the current user profile.
+   */
+  profile(): Promise<OidcProfileWithCustomClaims<TCustomClaims>>;
+  /**
+   * Gets a token.
+   */
+  token(): Promise<string>;
+  /**
+   * Extends the headers of the provided request.
+   */
+  extendHeaders(req: OidcRequest): void;
+}
+
+export interface PiralOidcApi<TCustomClaims extends {} = {}> {
   /**
    * Gets the currently valid access token, if any.
    */
   getAccessToken(): Promise<string | undefined>;
+
+  getProfile(): Promise<OidcProfileWithCustomClaims<TCustomClaims>>;
+}
+
+declare module 'piral-core/lib/types/custom' {
+  interface PiletCustomApi<TCustomClaims extends {} = {}> extends PiralOidcApi<TCustomClaims> {}
```

Comment:
(Just to be on the same page - that's I guess the reason for adding the default type parameter `={}`, however, it will always select the default one resulting in `PiralOidcApi<{}>`, i.e., removing the custom information.)

Reply:
Did you have a particular module in mind to place the custom interface on? Keep it in `piral-core/custom...` section, or a new module entirely just for this plugin? I'm not sure if any other plugins have their own "custom" area for declaration merging that I can copy the pattern of.

Reply:
No just include it in here (referred from a Piral instance `piral-oidc/lib/types`).

Other plugins make it similar (e.g., https://github.com/smapiot/piral/blob/develop/src/packages/piral-menu/src/types.ts#L87) and mention this in the README (e.g., https://github.com/smapiot/piral/tree/develop/src/packages/piral-menu#customizing).

Reply:
Potentially we will improve customizations / interface merging a lot soon - if we decide to ship bundled `d.ts` in `index.d.ts` of the packages, e.g., such that

```ts
declare module 'piral-core' {
  interface PiralCustomEventMap {
    // extend it
  }
}
```

would work, instead of `declare module 'piral-core/lib/types/custom'`. This reduces the required knowledge from "where is it exactly" to "what's the name".

Reply:
Updated to use piral-oidc/lib/types `PiralCustomOidcProfile`.